### PR TITLE
Fix #1262: Add correct implementation of NumberToString

### LINKED ISSUE #1262: Wrong constant-folding of the number to string conversion
The constant-folder, or some other parts of GCC, wrongly handles the conversion from number to string. This happens both with Simple Opts and Advanced Opts. The following snippet reproduces the problem:
```javascript
function foldNumberToString() {
  alert('A number: ' + 1.2323919403474454e+21.toString());
  alert('A number: ' + 1.2323919403474454e+21);
}
foldNumberToString();
```
The first call is not constant-folded, and outputs the correct results, i.e., `1.2323919403474454e+21`. The second line, however, is constant-folded into
```javascript
  alert("A number: 1.2323919403474454E21");
```
which seems to indicate that a `.toString()` was used on the JVM to convert the number into a string.

ECMAScript [specifies pretty precisely](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-tostring-applied-to-the-number-type) how a number must be converted into a string, and the algorithm does *not* match what the JVM does.

### PR DESCRIPTION
**Pull request by paullepoulpe**
[Review on Reviewable](https://reviewable.io/reviews/google/closure-compiler/1267)

### CODE REVIEW

**Comment by Dominator008** [comment]
@MatrixFrog Will it be easyto merge this internally?

**Comment by concavelenz** [comment]
Ben, what is happening with this?

**Comment by blickly** [comment]
Sorry to spend so long on this.
I think that this is good in terms of functionality. The only issue I'm having are with tracking down the licensing of all the different files.

Is there any way we can use the slightly older versions of these files that still had the MPL v1 license for consistency with the rest of our Rhino codebase?
Also it would help if the file you wrote yourself also includes a license header (similar to the other files in that directory).

**Comment by paullepoulpe** [comment]
The license version change was actually the last modification made to those files.

**Comment by blickly** [comment]
Cool, thanks for the update!

File: test/com/google/javascript/rhino/dtoa/NumberToStringTest.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ * MPL, indicate your decision by deleting the provisions above and replacing
+ * them with the notice and other provisions required by the GPL. If you do
+ * not delete the provisions above, a recipient may use your version of this
+ * file under either the MPL or the GPL.
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+package com.google.javascript.rhino.dtoa;
+
+import com.google.javascript.rhino.dtoa.DToA;
```
Comment (blickly) [inline test/com/google/javascript/rhino/dtoa/NumberToStringTest.java]: I think this import is superfluous.

File: test/com/google/javascript/rhino/dtoa/NumberToStringTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+import junit.framework.TestCase;
+
+public class NumberToStringTest extends TestCase {
+
+  public NumberToStringTest() {
```
Comment (blickly) [inline test/com/google/javascript/rhino/dtoa/NumberToStringTest.java]: Same with this constructor.

**Comment by paullepoulpe** [comment]
Bump

**Pull request closed without merging**